### ISSUE TITLE
Clarity on supported models

### ISSUE THREAD

**Issue by ammunk** [f-dangel/backpack#82]
Thanks for creating an honestly amazing package for speeding up batched gradient calculation!

# Summary

I have a few clarity questions regarding what modules are and are not supported and in which cases.

# Specifics

### Expecting sequences
In [supported models](https://docs.backpack.pt/en/master/supported-layers.html) is it stated that backpack expect models to be sequences (`nn.Sequential`). However, in the [ResNet](https://docs.backpack.pt/en/master/use_cases/example_first_order_resnet.html#sphx-glr-use-cases-example-first-order-resnet-py) example this is not the case.

- My question is then, does backpack only expect sequences of modules for second order extensions?

### For first-order extensions support any modules (without parameters)
- Are parameters defines as "learnable" parameters? I.e. would modules `nn.LeakyReLU` work? They have a parameters but they are not trainable.

### Are nested modules supported?

Say I define 

``` 
class Child(nn.Module):
    ...

class Parent(nn.Module):
    def __init__(self):
        super(Parent,self).__init__()
        self.child = Child()
```

Are the parameters in `Child` properly tracked by backpack?

**Comment by fKunstner** [comment]
Hi Andreas,

Thanks for reaching out! Answers below, hope those help. I'll add those clarification to the docs. 
Glad you appreciate the package :)

---

> **Expecting sequences**
> does backpack only expect sequences of modules for second order extensions?

Yes, the "strict" constraints on the architecture apply to second-order extensions. 
The first-order ones are more flexible, as in the [ResNet example](https://docs.backpack.pt/en/master/use_cases/example_first_order_resnet.html#sphx-glr-use-cases-example-first-order-resnet-py), as we can piggyback on PyTorch.

> **For first-order extensions support any modules (without parameters)**
> Are parameters "learnable" parameters? 

Yes. Maybe another way to phrase this is that anything that does not `requires_grad` is fine for first-order extensions;
the gradient information for the layers with learnable parameters will be propagated and extracted. 
The gradients of the individual losses with respect to the parameters of the `LeakyReLU` is not (yet) supported though.

> **Are nested modules supported?**

Yes. For first-order extensions, if the child-most layers are instances of `nn.Linear` or `nn.Conv2D`, backpack should find them and extract the first-order extensions. The class `MyFirstResNet` in the [ResNet example](https://docs.backpack.pt/en/master/use_cases/example_first_order_resnet.html#sphx-glr-use-cases-example-first-order-resnet-py) works that way. 

We don't have extensive tests on non-sequential architectures though, so we might have missed something. Let us know if you run into issues.

**Comment by ammunk** [comment]
Hi Frederik,

Thank you very much for getting back for me so quickly. That certainly made it clear! :)

I haven't encountered any issues actually and my tests so far shows that it works the way I expect. I just wanted to make sure I wasn't making any wrong assumptions only to encounter serious issues later.

Again, thanks for a great package. I'll close this.
